src/Microsoft.AspNet.OData.Shared/Query/ODataPathQueryBuilder.cs

```diff
@@ -0,0 +1,315 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using Microsoft.OData.Edm;
+using Microsoft.OData.UriParser;
+
+namespace Microsoft.AspNet.OData.Query
+{
+
+    /// <summary>
+    /// This class transforms an <see cref="IQueryable"/> based on
+    /// the key and property accesses in the segments of an <see cref="ODataPath"/>
+    /// </summary>
+    internal class ODataPathQueryBuilder
+    {
+        readonly IQueryable source;
+        readonly IEdmModel model;
+        readonly Routing.ODataPath path;
+
+        /// <summary>
+        /// Creates an instance of <see cref="ODataPathQueryBuilder"/>
+        /// </summary>
+        /// <param name="source">The original <see cref="IQueryable"/> to be transformed</param>
+        /// <param name="model">The <see cref="IEdmModel"/></param>
+        /// <param name="path">The <see cref="ODataPath"/> based on which to transform the <see cref="source"/></param>
+        public ODataPathQueryBuilder(IQueryable source, IEdmModel model, Routing.ODataPath path)
+        {
+            this.source = source;
+            this.model = model;
+            this.path = path;
+        }
+
+        /// <summary>
+        /// Transforms the source <see cref="IQueryable"/> based on the sequence of path segments in the <see cref="ODataPath"/>
+        /// </summary>
+        /// <returns>The result of the query transformtions or null if the path contained unsupported segments</returns>
+        public ODataPathQueryResult BuildQuery()
+        {
+            var result = new ODataPathQueryResult();
+
+            var segments = path.Segments;
+            // assume first segment is entitySet
+            var firstSegment = segments.FirstOrDefault() as EntitySetSegment;
```

Comment:
Thanks for catching this, actually `firstSegment` variable is not used. The code works fine even if first segment is a singleton since query expressions are applied starting from the second. I will remove this statement.